Add AMX RTB adapter

nickjacob opened the pull request:
Adding AMX RTB adapter

Server uses standard OpenRTB 2.5. This adapter adds support for overriding the publisher id using `imp.ext.amx.tagId`, although this is optional. The parameter is named to match the bid param of [our prebid.js adapter](https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/amx-rtb.md)

Maintainer email: prebid@amxrtb.com

bretg commented:
docs PR https://github.com/prebid/prebid.github.io/pull/2437

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+
+// MakeRequests creates AMX adapter requests
+func (adapter *AmxAdapter) MakeRequests(request *openrtb.BidRequest, req *adapters.ExtraRequestInfo) (reqsBidder []*adapters.RequestData, errs []error) {
+	publisherID := getPublisherID(request.Imp)
```
With the other suggested change, this would become:
```
if publisherID, ok := getPublisherID(request.Imp); ok {
	if request.App != nil {
		request.App.Publisher = ensurePublisherWithID(request.App.Publisher, publisherID)
	}
	if request.Site != nil {
		request.Site.Publisher = ensurePublisherWithID(request.Site.Publisher, publisherID)
	}
}
```

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+				return &params.Bidder.PublisherID
+			}
+		} else {
+			fmt.Printf("unable to decode imp.ext.bidder: %v", err)
```
There is no consequence to this print statement, neither in the response or the in the log. The chance of this happening is also likely 0% due to preprocessing by the "core" logic. I recommend removing the print statement since the publisher id is optional for your adapter.

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+		}
+	}
+
+	encoded, err := json.Marshal(request)
```
Nitpick: In Go, it's common to remove the line break after the `json.Marshal` call to group it with the error handling code.

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	reqBidder := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     adapter.endpoint + "?v=pbs1.0",
```
Please consider adding this query parameter safely using Go's url methods within the `NewAmxBidder` method.

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	if http.StatusOK != response.StatusCode {
+		return nil, []error{&errortypes.BadInput{
```
Since this error would likely mean the server responded back with a bad response, it's common to use `BadServerResponse` here instead.

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+	return "", &errortypes.BadInput{
+		Message: fmt.Sprintf("No impression for: %s", impID),
+	}
+}
```
Does your bid response include the bid type? Although the code you have here is common in the project, there's been issues reported (see https://github.com/prebid/prebid-server/issues/862) that this approach isn't always reliable for multiformat impressions. Is there enough information to detect the type based on the bid response?

SyntaxNode commented on adapters/amx/usersync.go:
```diff
@@ -0,0 +1,13 @@
+
+// NewAmxSyncer produces an AMX RTB usersyncer
+func NewAmxSyncer(temp *template.Template) usersync.Usersyncer {
+	return adapters.NewSyncer("amx", 737, temp, adapters.SyncTypeRedirect)
```
GVL ID is registered to `Monet Engine Inc`. Is that correct for AMX?

SyntaxNode commented on adapters/amx/usersync_test.go:
```diff
@@ -0,0 +1,24 @@
+	assert.Equal(t, testSyncEndpoint, syncInfo.URL)
+	assert.Equal(t, false, syncInfo.SupportCORS)
+	assert.EqualValues(t, 737, syncer.GDPRVendorID())
+}
```
Please follow the format form the other syncer tests.

SyntaxNode commented on usersync/usersyncers/syncer.go:
```diff
@@ -99,6 +100,7 @@ func NewSyncerMap(cfg *config.Configuration) map[openrtb_ext.BidderName]usersync
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderAdtelligent, adtelligent.NewAdtelligentSyncer)
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderAdvangelists, advangelists.NewAdvangelistsSyncer)
 	insertIntoMap(cfg, syncers, openrtb_ext.BidderAJA, aja.NewAJASyncer)
+	insertIntoMap(cfg, syncers, openrtb_ext.BidderAmx, amx.NewAmxSyncer)
```
Per Go naming convention, this should be `BidderAMX`, `NewAMXSyncer`, etc...

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+type amxParams struct {
+	PublisherID string `json:"tagId,omitempty"`
```
Please move `amxParams` to the `openrtb_ext` package in the file openrtb_ext/imp_amx.go with the name `ExtImpAmx` to match our standard convention.

SyntaxNode reviewed (commented):
Thank you for your contribution. Very glad to see the docs PR already open. :)

nickjacob commented:
Sorry for some reason github won't let me resolve the comments or reply 
here—but thanks for the great CR!

I made the changes you noted. Our bids do always have `startdelay` in 
bid.ext when video, so I used that instead to determine the type.

For the GVL—737 is correct thanks

I renamed everything to AMX (from Amx). Let me know if there's anything 
else

Thanks!

On 26 Oct 2020, at 16:15, Scott Kay wrote:

> @SyntaxNode commented on this pull request.
>
> Thank you for your contribution. Very glad to see the docs PR already 
> open. :)
>
>> +	"net/http"
> +	"strings"
> +)
> +
> +const vastImpressionFormat = "<Impression><![CDATA[%s]></Impression>"
> +const vastSearchPoint = "</Impression>"
> +
> +// AmxAdapter is the AMX bid adapter
> +type AmxAdapter struct {
> +	endpoint string
> +}
> +
> +// NewAmxBidder creates an AmxAdapter
> +func NewAmxBidder(endpoint string) *AmxAdapter {
> +	return &AmxAdapter{endpoint: endpoint}
> +}
>
> FYI: We'll update the `NewAmxBidder` for you in the coming weeks with 
> the https://github.com/prebid/prebid-server/pull/1532 PR.
>
>> +
> +func getPublisherID(imps []openrtb.Imp) *string {
> +	for _, imp := range imps {
> +		paramsSource := (*json.RawMessage)(&imp.Ext)
> +		var params amxExt
> +		if err := json.Unmarshal(*paramsSource, &params); err == nil {
> +			if params.Bidder.PublisherID != "" {
> +				return &params.Bidder.PublisherID
> +			}
> +		} else {
> +			fmt.Printf("unable to decode imp.ext.bidder: %v", err)
> +		}
> +	}
> +
> +	return nil
> +}
>
> Logic looks good, but the use of pointers is largely unnecessary. It's 
> common in Go to return a second boolean to indiciate if the string was 
> found or not rather than using nil. Something like this would be more 
> idiomatic Go.
>
> ```
> func getPublisherID(imps []openrtb.Imp) (string, bool) {
> 	for _, imp := range imps {
> 		var params amxExt
> 		if err := json.Unmarshal(imp.Ext, &params); err == nil {
> 			if params.Bidder.PublisherID != "" {
> 				return params.Bidder.PublisherID, true
> 			}
> 		} else {
> 			fmt.Printf("unable to decode imp.ext.bidder: %v", err)
> 		}
> 	}
>
> 	return "", false
> }
> ```
>
>> +	}
> +
> +	return nil
> +}
> +
> +func ensurePublisherWithID(pub *openrtb.Publisher, publisherID 
> string) *openrtb.Publisher {
> +	if pub == nil {
> +		pub = &openrtb.Publisher{}
> +	}
> +	pub.ID = publisherID
> +	return pub
> +}
> +
> +// MakeRequests creates AMX adapter requests
> +func (adapter *AmxAdapter) MakeRequests(request *openrtb.BidRequest, 
> req *adapters.ExtraRequestInfo) (reqsBidder []*adapters.RequestData, 
> errs []error) {
> +	publisherID := getPublisherID(request.Imp)
>
> With the other suggested change, this would become:
> ```
> if publisherID, ok := getPublisherID(request.Imp); ok {
> 	if request.App != nil {
> 		request.App.Publisher = ensurePublisherWithID(request.App.Publisher, 
> publisherID)
> 	}
> 	if request.Site != nil {
> 		request.Site.Publisher = 
> ensurePublisherWithID(request.Site.Publisher, publisherID)
> 	}
> }
> ```
>
>> +}
> +
> +type amxParams struct {
> +	PublisherID string `json:"tagId,omitempty"`
> +}
> +
> +func getPublisherID(imps []openrtb.Imp) *string {
> +	for _, imp := range imps {
> +		paramsSource := (*json.RawMessage)(&imp.Ext)
> +		var params amxExt
> +		if err := json.Unmarshal(*paramsSource, &params); err == nil {
> +			if params.Bidder.PublisherID != "" {
> +				return &params.Bidder.PublisherID
> +			}
> +		} else {
> +			fmt.Printf("unable to decode imp.ext.bidder: %v", err)
>
> There is no consequence to this print statement, neither in the 
> response or the in the log. The chance of this happening is also 
> likely 0% due to preprocessing by the "core" logic. I recommend 
> removing the print statement since the publisher id is optional for 
> your adapter.
>
>> +	return pub
> +}
> +
> +// MakeRequests creates AMX adapter requests
> +func (adapter *AmxAdapter) MakeRequests(request *openrtb.BidRequest, 
> req *adapters.ExtraRequestInfo) (reqsBidder []*adapters.RequestData, 
> errs []error) {
> +	publisherID := getPublisherID(request.Imp)
> +	if publisherID != nil {
> +		if request.App != nil {
> +			request.App.Publisher = 
> ensurePublisherWithID(request.App.Publisher, *publisherID)
> +		}
> +		if request.Site != nil {
> +			request.Site.Publisher = 
> ensurePublisherWithID(request.Site.Publisher, *publisherID)
> +		}
> +	}
> +
> +	encoded, err := json.Marshal(request)
>
> Nitpick: In Go, it's common to remove the line break after the 
> `json.Marshal` call to group it with the error handling code.
>
>> +			request.Site.Publisher = 
>> ensurePublisherWithID(request.Site.Publisher, *publisherID)
> +		}
> +	}
> +
> +	encoded, err := json.Marshal(request)
> +
> +	if err != nil {
> +		errs = append(errs, err)
> +		return nil, errs
> +	}
> +
> +	headers := http.Header{}
> +	headers.Add("Content-Type", "application/json;charset=utf-8")
> +	reqBidder := &adapters.RequestData{
> +		Method:  "POST",
> +		Uri:     adapter.endpoint + "?v=pbs1.0",
>
> Please consider adding this query parameter safely using Go's url 
> methods within the `NewAmxBidder` method.
>
>> +// MakeBids will parse the bids from the AMX server
> +func (adapter *AmxAdapter) MakeBids(request *openrtb.BidRequest, 
> externalRequest *adapters.RequestData, response 
> *adapters.ResponseData) (*adapters.BidderResponse, []error) {
> +	var errs []error
> +
> +	if http.StatusNoContent == response.StatusCode {
> +		return nil, nil
> +	}
> +
> +	if http.StatusBadRequest == response.StatusCode {
> +		return nil, []error{&errortypes.BadInput{
> +			Message: fmt.Sprint("Invalid request: 400"),
> +		}}
> +	}
> +
> +	if http.StatusOK != response.StatusCode {
> +		return nil, []error{&errortypes.BadInput{
>
> Since this error would likely mean the server responded back with a 
> bad response, it's common to use `BadServerResponse` here instead.
>
>> +
> +func getMediaTypeForImp(impID string, imps []openrtb.Imp) 
> (openrtb_ext.BidType, error) {
> +	mediaType := openrtb_ext.BidTypeBanner
> +	for _, imp := range imps {
> +		if imp.ID == impID {
> +			if imp.Banner == nil && imp.Video != nil {
> +				mediaType = openrtb_ext.BidTypeVideo
> +			}
> +			return mediaType, nil
> +		}
> +	}
> +
> +	return "", &errortypes.BadInput{
> +		Message: fmt.Sprintf("No impression for: %s", impID),
> +	}
> +}
>
> Does your bid response include the bid type? Although the code you 
> have here is common in the project, there's been issues reported (see 
> https://github.com/prebid/prebid-server/issues/862) that this approach 
> isn't always reliable for multiformat impressions. Is there enough 
> information to detect the type based on the bid response?
>
>> @@ -0,0 +1,13 @@
> +package amx
> +
> +import (
> +	"text/template"
> +
> +	"github.com/prebid/prebid-server/adapters"
> +	"github.com/prebid/prebid-server/usersync"
> +)
> +
> +// NewAmxSyncer produces an AMX RTB usersyncer
> +func NewAmxSyncer(temp *template.Template) usersync.Usersyncer {
> +	return adapters.NewSyncer("amx", 737, temp, 
> adapters.SyncTypeRedirect)
>
> GVL ID is registered to `Monet Engine Inc`. Is that correct for AMX?
>
>> +)
> +
> +const (
> +	testSyncEndpoint = "http://pbs.amxrtb.com/cchain/0"
> +)
> +
> +func TestDmxSyncer(t *testing.T) {
> +	temp := 
> template.Must(template.New("sync-template").Parse(testSyncEndpoint))
> +	syncer := NewAmxSyncer(temp)
> +	syncInfo, err := syncer.GetUsersyncInfo(privacy.Policies{})
> +	assert.NoError(t, err)
> +	assert.Equal(t, "redirect", syncInfo.Type)
> +	assert.Equal(t, testSyncEndpoint, syncInfo.URL)
> +	assert.Equal(t, false, syncInfo.SupportCORS)
> +	assert.EqualValues(t, 737, syncer.GDPRVendorID())
> +}
>
> Please follow the format form the other syncer tests.
>
>> @@ -99,6 +100,7 @@ func NewSyncerMap(cfg *config.Configuration) 
>> map[openrtb_ext.BidderName]usersync
>  	insertIntoMap(cfg, syncers, openrtb_ext.BidderAdtelligent, 
> adtelligent.NewAdtelligentSyncer)
>  	insertIntoMap(cfg, syncers, openrtb_ext.BidderAdvangelists, 
> advangelists.NewAdvangelistsSyncer)
>  	insertIntoMap(cfg, syncers, openrtb_ext.BidderAJA, aja.NewAJASyncer)
> +	insertIntoMap(cfg, syncers, openrtb_ext.BidderAmx, amx.NewAmxSyncer)
>
> Per Go naming convention, this should be `BidderAMX`, `NewAMXSyncer`, 
> etc...
>
>> +// AmxAdapter is the AMX bid adapter
> +type AmxAdapter struct {
> +	endpoint string
> +}
> +
> +// NewAmxBidder creates an AmxAdapter
> +func NewAmxBidder(endpoint string) *AmxAdapter {
> +	return &AmxAdapter{endpoint: endpoint}
> +}
> +
> +type amxExt struct {
> +	Bidder amxParams `json:"bidder"`
> +}
> +
> +type amxParams struct {
> +	PublisherID string `json:"tagId,omitempty"`
>
> Please move `amxParams` to the `openrtb_ext` package in the file 
> openrtb_ext/imp_amx.go with the name `ExtImpAmx` to match our standard 
> convention.
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/prebid/prebid-server/pull/1549#pullrequestreview-517126367

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,158 @@
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			var bidExt amxBidExt
+			if err := json.Unmarshal(bid.Ext, &bidExt); err == nil {
```
There's a problem here because if `bid.Ext` is empty the `json.Unmarshal` call will return an error and not produce a bid response. This doesn't fail the exemplary tests because you're missing `expectedBidResponses` from those files.

This will need to be changed to handle a missing `bid.Ext`. Perhaps something like:
```
for _, sb := range bidResp.SeatBid {
		for _, bid := range sb.Bid {

			bidExt, bidExtErr := getBidExt(bid.Ext)
			if bidExtErr != nil {
				errs = append(errs, bidExtErr)
				continue
			}

			bidType := getMediaTypeForBid(bidExt)

			b := &adapters.TypedBid{
				Bid:     &bid,
				BidType: bidType,
			}
			if b.BidType == openrtb_ext.BidTypeVideo {
				b.Bid.AdM = interpolateImpressions(bid, bidExt)
			}
			bidResponse.Bids = append(bidResponse.Bids, b)

		}
	}
```

```
func getBidExt(ext json.RawMessage) (amxBidExt, error) {
	if len(ext) == 0 {
		return amxBidExt{}, nil
	}

	var bidExt amxBidExt
	err := json.Unmarshal(ext, &bidExt)

	return bidExt, err
}
```

SyntaxNode commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,354 @@
+	if body.Site.Publisher.ID != "publisher_id_example" {
+		t.Errorf("invalid publisher id: %s", body.Site.Publisher.ID)
+	}
+}
```
Please consider using testify assert statements in your tests. Not strictly required. A bit easier to read though and you don't need to be concerned with edge cases.

```
assert.NoError(t, actualAdapterRequests, "request")
assert.Empty(t, err, "request errors")
...
assert.Equal(t, "publisher_id_example", body.Site.Publisher.ID, "publisher")
```

SyntaxNode commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,354 @@
+		},
+		App: &openrtb.App{ID: "cansanuabnua", Publisher: &openrtb.Publisher{ID: "whatever"}},
+		ID:  "1234",
+	}
```
It's not valid for a BidRequest to contain both a `Site` and `App` section. This will be caught earlier on by the framework. For this test, please only use the `App`.

SyntaxNode commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,354 @@
+            "w": 300,
+            "ext": {
+							"himp": ["https://example.com/imp-tracker/pixel.gif?param=1&param2=2"],
+							"startdelay": 0
```
Please fix the whitespace on these lines for consistency.

SyntaxNode commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,354 @@
+		},
+		User: &openrtb.User{Ext: json.RawMessage(`{"eids": [{
+						"source": "adserver.org",
+							"uids": [{
```
Same here. Please use consistent whitespace for the json blobs.

SyntaxNode commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,354 @@
+		t.Errorf("String was not found")
+	}
+
+	if strings.Index(data, "example.com/pixel.png") == -1 {
```
assert.Contains(t, data, "example.com/pixel.png", "interpolated into vast")

SyntaxNode commented on adapters/amx/amxtest/exemplary/app.simple.json:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+  ]
+}
```
Please add expectedBidResponses to the json test files.

SyntaxNode reviewed (commented):
Thank you for the quick update from our feedback. I took a closer look into the tests and found a few items to address.

nickjacob commented:
Thanks @SyntaxNode  -- I made the requested CR changes; refactored to use assert which did really improve the tests, adding some more test coverage and "supplemental" JSON tests, expectedBidResponse & expectedMakeBidsErrors in JSON tests

mansinahar commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,335 @@
+	userExt := getUserExt(body.User)
+	assert.NotNil(t, userExt, "the generated user.ext is invalid")
+
+	assert.Len(t, userExt.Eids, 2)
```
Not exactly sure what exactly this test is testing because I don't see any modifications around user ext in your `MakeRequests` function. This to me looks more like making sure that user exts are not getting modified by `MakeRequests`. Is that correct? 

Also, I'd suggest adding these tests in this file as part of the JSON tests rather than unit tests as that will help remove a lot of setup for your tests which is covered by `RunJSONBidderTest` and will also provide consistency in tests with other adapters. In fact some of these test cases are already covered with the JSON tests.

mansinahar commented on static/bidder-params/amx.json:
```diff
@@ -0,0 +1,12 @@
+  "description": "A schema to validate params accepted by the AMX adapter",
+  "type": "object",
+  "properties": {
+    "publisher_id" : {
```
Bidder params should have the params expected by your adapter which in your case should be `tagid`. You will have to update the `adapters/amx/amxtest/params/race/display.json` and `adapters/amx/amxtest/params/race/video.json` file accordingly as well.

nickjacob commented:
Thanks @mansinahar -- I made those changes & rebased against master

mansinahar commented on adapters/amx/amx_test.go:
```diff
@@ -0,0 +1,335 @@
+
+func TestJsonSamples(t *testing.T) {
+	t.Logf("TESTING JSON SAMPLES!!")
+	adapterstest.RunJSONBidderTest(t, "amxtest", new(AMXAdapter))
```
@nickjacob Could you please address this? looks like this was missed in the update. Thanks!

nickjacob replied:
sorry about that missed this -- just pushed another commit + with the table-driven test for the publisher ID logic

mansinahar reviewed (commented):
Please add a `params_test.go` file as well to test the bidder params. For reference you can check this one for 33across: https://github.com/prebid/prebid-server/blob/master/adapters/33across/params_test.go

SyntaxNode commented on adapters/amx/amx.go:
```diff
@@ -0,0 +1,197 @@
+		return nil
+	}
+
+	qs, _ := url.ParseQuery(endpointURL.RawQuery)
```
Please handle the error from `url.ParseQuery` in a similar manner as you did with `url.Parse`.

nickjacob commented:
added params_test.go, improved testing of interpolation function + better error handling in `NewAMXBidder`. Also added default sync endpoint

also rebased off of master just to be safe

SyntaxNode commented:
@nickjacob Can you please address my comment on the docs PR?

SyntaxNode reviewed (approved):
I'm good to merge this in. The docs PR is open with one remaining comment, but I don't see a reason for that to be a blocker here.

nickjacob commented:
@SyntaxNode  -- for the docs PR; we don't want to have all those same parameters for PBS that we have in PBJS (e.g. testMode is the same as test: 1, so not necessary). Should I make a second table in the doc that describes the PBS parameters separate from PBJS?

SyntaxNode commented:
> @SyntaxNode -- for the docs PR; we don't want to have all those same parameters for PBS that we have in PBJS (e.g. testMode is the same as test: 1, so not necessary). Should I make a second table in the doc that describes the PBS parameters separate from PBJS?

Yes, please. The `testMode` makes sense. 

Before I git the merge button... what about `tagId` vs `publisher_id` and the `adUnitId`?

nickjacob commented:
I ended up changing the parameter here to be "tagId" to match "tagId" in PBJS. adUnitId is probably a good idea you're right. I'll add that now so the PBS version better matches PBJS

SyntaxNode commented:
> I ended up changing the parameter here to be "tagId" to match "tagId" in PBJS. adUnitId is probably a good idea you're right. I'll add that now so the PBS version better matches PBJS

Ah, right. Sorry. Forgot about the `tagId` change was already made. I'll hold off on merging for the `adUnitId` then. Thank you.
